File: python/src/azoth/eos/reference/tv_flash.py
"""``eos.tv_flash`` - the pressure a mixture reaches at a given temperature and volume.

Spec: ``specs/models/eos/tv_flash.toml``

An outer quasi-Newton on pressure, over the molar volume assembled from the phase split
at a trial pressure (:mod:`azoth.eos.reference.pt_flash`). The volume falls monotonically
with pressure at a fixed temperature. The loop lives in
:mod:`azoth.eos.reference._flash_property`, shared with the other flash-breadth models.
"""

from __future__ import annotations

from azoth import _models_gen
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import TvFlashResult
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.mixture import Mixture
from azoth.eos.reference._flash_property import distinct_warnings, solve_pressure
from azoth.eos.reference.molar_enthalpy_entropy import IdealGasModel
from azoth.eos.reference.pr_molar_volume import MOLAR_GAS_CONSTANT

MODEL_ID = "eos.tv_flash"


def tv_flash(
    mixture: Mixture,
    ideal_gas: IdealGasModel,
    T: Q,
    V: Q,
    z: list[float],
) -> TvFlashResult:
    """The pressure at which a mixture has a given molar volume at a temperature.

    Raises:
        OutOfRangeError: if ``T`` or ``V`` is not positive.
        SolverNotConvergedError: if the iteration reaches its cap.
    """
    spec = _models_gen.model(MODEL_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []

    t_si = input_to_si(spec, "T", T)
    v_si = input_to_si(spec, "V", V)
    apply_checks(checks.on_input, {"T": t_si, "V": v_si}.get, warnings)

    start = MOLAR_GAS_CONSTANT * t_si / v_si
    solved = solve_pressure(mixture, ideal_gas, t_si, v_si, list(z), "v", spec["algorithm"], start)
    warnings.extend(solved["warnings"])
    flash = solved["state"]["flash"]

    apply_checks(
        checks.derived,
        lambda name: solved["P"] if name == "P" else None,
        warnings,
    )

    return TvFlashResult(
        P=from_si(solved["P"], "Pa"),
        beta=flash.beta,
        x=tuple(flash.x),
        y=tuple(flash.y),
        k=tuple(flash.k),
        phase=flash.phase,
        z_liquid=flash.z_liquid,
        z_vapour=flash.z_vapour,
        iterations=solved["iterations"],
        residual=solved["residual"],
        warnings=distinct_warnings(warnings),
    )

File: crates/azoth-eos/src/tv_flash.rs
//! `eos.tv_flash` - the pressure a mixture reaches at a given temperature and molar volume.
//!
//! Spec: `specs/models/eos/tv_flash.toml`
//!
//! An outer quasi-Newton on pressure, over the molar volume assembled from the phase split
//! at a trial pressure ([`crate::pt_flash`]) and each phase's compressibility. The volume
//! falls monotonically with pressure at a fixed temperature, which is what makes the
//! inversion well posed. The loop lives in [`crate::flash_property`].

use azoth_core::units::{MolarVolume, ThermodynamicTemperature, pascals};
use azoth_core::{Result, apply_checks};

use crate::algorithm_of;
use crate::flash_property::{Property, solve_pressure};
use crate::mixture::Mixture;
use crate::model_gen;
use crate::molar_enthalpy_entropy::IdealGasModel;
use crate::pr_molar_volume::MOLAR_GAS_CONSTANT;
use crate::results::TvFlashResult;

/// The pressure at which a mixture has a given molar volume at a temperature.
///
/// # Errors
/// * [`azoth_core::AzothError::OutOfRange`] if `t` or `v` is not positive.
/// * [`azoth_core::AzothError::SolverNotConverged`] if the iteration reaches its cap.
///
/// # Example
/// ```
/// use azoth_core::units::{cubic_meters_per_mole, kelvins};
/// use azoth_eos::{databank, tv_flash};
///
/// let (mixture, ideal_gas) = databank::mixture_of(&["methane", "n-butane"], None)
///     .expect("the pair resolves");
/// let r = tv_flash::tv_flash(
///     &mixture,
///     &ideal_gas,
///     kelvins(300.0),
///     cubic_meters_per_mole(1.9610505645240935e-3),
///     &[0.6, 0.4],
/// )?;
/// assert!((r.pressure.value - 1.0e6).abs() < 10.0);
/// # Ok::<(), azoth_core::AzothError>(())
/// ```
pub fn tv_flash(
    mixture: &Mixture,
    ideal_gas: &IdealGasModel,
    t: ThermodynamicTemperature,
    v: MolarVolume,
    z: &[f64],
) -> Result<TvFlashResult> {
    let spec = &model_gen::TV_FLASH_SPEC;
    let mut warnings = Vec::new();

    apply_checks(
        spec.input_checks(),
        |quantity| match quantity {
            "T" => Some(t.value),
            "V" => Some(v.value),
            _ => None,
        },
        &mut warnings,
    )?;

    let algorithm = algorithm_of(spec)?;
    // The ideal-gas pressure for this temperature and volume: a first guess within an
    // order of magnitude of the true pressure, for either a gas or a condensed phase.
    let start = MOLAR_GAS_CONSTANT * t.value / v.value;
    let solved = solve_pressure(
        mixture,
        ideal_gas,
        t,
        v.value,
        z,
        Property::Volume,
        algorithm,
        start,
    )?;

    apply_checks(
        spec.derived_checks(),
        |quantity| (quantity == "P").then_some(solved.pressure),
        &mut warnings,
    )?;

    Ok(TvFlashResult {
        pressure: pascals(solved.pressure),
        beta: solved.flash.beta,
        x: solved.flash.x,
        y: solved.flash.y,
        k: solved.flash.k,
        phase: solved.flash.phase,
        z_liquid: solved.flash.z_liquid,
        z_vapour: solved.flash.z_vapour,
        iterations: solved.iterations,
        residual: solved.residual,
        warnings: solved.warnings,
    })
}
